README
Gtk2::Unique
=========

Perl bindings for the C library "libunique" that provides a mechanism for
writing single instance applications. If you launch a single instance
application twice, the second instance will either just quit or will send a
message to the running instance.

Unique makes it easy to write this kind of applications, by providing a base
class, taking care of all the IPC machinery needed to send messages to a running
instance, and also handling the startup notification side.

For more information about libunique refer to the library's web site:

	http://live.gnome.org/LibUnique


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

	perldoc Gtk2::Unique

Also feel free to take a look at the examples in the folder "examples/" and to
play around with them.

For any kind of help or support simply report a bug or join us in the IRC
channel #gtk2-perl on freenode.


COPYRIGHT AND LICENCE

Copyright (C) 2009 Emmanuel Rodriguez

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for Gtk2::Unique

0.04 Tue Nov  3 22:27:54 CET 2009
	The signal Gtk2::UniqueApp::message-received now receives command name
	instaed of the command id.
	The macro UNIQUE_CHECK_VERSION will be fixed in 1.1.0.


0.03 Tue Oct 20 20:21:27 CEST 2009
	Add the version check macros to the bindings.
	Fix the unit tests (now using the bacon backend).
	Documentation updates.


0.02 Sun Jul 12 16:28:57 CEST 2009
	More perlish API with send_message_by_name().


0.01 Thu Apr  2 15:44:36 CEST 2009
	First version, released on an unsuspecting world.


t/unit-tests.pl
#!/usr/bin/perl

use strict;
use warnings;

use Glib qw(TRUE FALSE);
use Gtk2 '-init';
use Gtk2::Unique;
use Data::Dumper;

my $COMMAND_FOO = 1;
my $COMMAND_BAR = 2;


exit main();


sub main {
	
	my $app = Gtk2::UniqueApp->new(
		"org.example.UnitTets", undef,
		foo => $COMMAND_FOO,
		bar => $COMMAND_BAR,
	);


	if ($app->is_running) {
		die "Application is already running";
	}


	# Create the single application instance and wait for other requests
	my $window = create_application($app);
	Gtk2->main();
	
	return 0;
}


#
# Called when the application needs to be created. This happens when there's no
# other instance running.
#
sub create_application {
	my ($app) = @_;

	# Standard window and windgets
	my $window = Gtk2::Window->new();
	$window->set_title("Gtk2::Unique - Unit Tests");
	$window->set_size_request(480, 240);
	my $textview = Gtk2::TextView->new();
	my $scroll = Gtk2::ScrolledWindow->new();
	my $buffer = $textview->get_buffer;

	# Widget packing
	$scroll->add($textview);
	$window->add($scroll);
	$window->show_all();

	# Widget signals
	$window->signal_connect(delete_event => sub {
		Gtk2->main_quit();
		return TRUE;
	});

	# Listen for new commands
	$app->watch_window($window);
	$app->signal_connect('message-received' => sub {
		my ($app, $command, $message, $time) = @_;
		
		my $text = Dumper($message->get);
		$buffer->insert($buffer->get_end_iter, "$command: $text\n");
		
		# The command FOO will succeed while the command BAR will fail
		return $command == $COMMAND_FOO ? 'ok' : 'invalid';
	});

	return $window;
}


Makefile.PL
use strict;
use warnings;

use ExtUtils::MakeMaker;

use Cwd;
use File::Spec;

use Gtk2::CodeGen;
use Glib::MakeHelper;
use ExtUtils::Depends;
use ExtUtils::PkgConfig;

my $DEPENDS;
my @XS_FILES = ();

exit main();

sub main {
	
	# Create the build folder used by the code generation utilities
	mkdir 'build', 0777;
	
	# Find libunique
	my @typemaps = ();
	my @deps = ('Glib');
	my %pkgconfig;
	
	eval {
		%pkgconfig = ExtUtils::PkgConfig->find("unique-1.0");
		push @XS_FILES, <xs/*.xs>;
		push @typemaps, 'maps';
		push @deps, 'Gtk2';
	};
	if (my $error = $@) {
		warn "FAIL: ", $error;
		return;
	}
	
	$DEPENDS = ExtUtils::Depends->new('Gtk2::Unique', @deps);
	
	$DEPENDS->add_pm(
		File::Spec->catfile('lib', 'Gtk2', 'Unique.pm'),
		File::Spec->catfile('$(INST_LIBDIR)', 'Unique.pm'),
	);
	
	# Code generation
	Gtk2::CodeGen->parse_maps('unique', input => [ @typemaps ]);
	Gtk2::CodeGen->write_boot(
		ignore   => qr/^Gtk2::Unique$/,
		xs_files => [ @XS_FILES ],
	);
	
	
	$DEPENDS->set_inc($pkgconfig{cflags} . ' -I./build');
	$DEPENDS->set_libs($pkgconfig{libs});
	$DEPENDS->add_xs(@XS_FILES);
	$DEPENDS->add_typemaps(
		File::Spec->catfile(cwd(), 'build', 'unique.typemap'),
	);
	$DEPENDS->install(
		File::Spec->catfile('build', 'unique-autogen.h'),
		'unique-perl.h',
	);
	$DEPENDS->save_config(File::Spec->catfile('build', 'IFiles.pm'));
	

	# Create the Makefile
	WriteMakefile(
		AUTHOR        => 'Emmanuel Rodriguez <potyl@cpan.org>',
		NAME          => 'Gtk2::Unique',
		VERSION_FROM  => File::Spec->catfile('lib', 'Gtk2', 'Unique.pm'),
		ABSTRACT_FROM => File::Spec->catfile('lib', 'Gtk2', 'Unique.pm'),
		LICENSE       => 'perl, lgpl',

		PREREQ_PM     => {
			'Gtk2' => '1.161', # Other verions might work too (Ubuntu 8.04 has this version)
		},
		PREREQ_FATAL  => 1,
		
		XSPROTOARG    => '-noprototypes ',
		MAN3PODS      => {
			Glib::MakeHelper->do_pod_files(@XS_FILES),
		},
		
		META_MERGE => {
			repository => 'http://github.com/potyl/gtk2-unique/tree/master',
		},
		
		$DEPENDS->get_makefile_vars(),
		
		# Remove the build folder when doing "make clean"
		clean => {
			FILES => 'build',
		},
	);
	
	return 0;
}


sub MY::postamble {
	
	my $postamble = Glib::MakeHelper->postamble_clean();
	$postamble .= Glib::MakeHelper->postamble_docs_full(
		DEPENDS   => $DEPENDS,
		XS_FILES  => [ @XS_FILES ],
		COPYRIGHT => 'Copyright (C) 2009 by Emmanuel Rodriguez'
	);
	
	return $postamble;
}


examples/sample.pl
#!/usr/bin/perl

use strict;
use warnings;

use Glib qw(TRUE FALSE);
use Gtk2 '-init';
use Gtk2::Unique;
use Encode;
use Data::Dumper;

my $COMMAND_WRITE = 1;


exit main();


sub main {
	die "Usage: message\n" unless @ARGV;
	my ($text) = @ARGV;
	# If we want to pass UTF-8 text in the command line arguments
	$text = decode('UTF-8', $text);
	
	# As soon as we create the UniqueApp instance we either have the name we
	# requested ("org.mydomain.MyApplication", in the example) or we don't because
	# there already is an application using the same name.
	my $app = Gtk2::UniqueApp->new(
		"org.example.Sample", undef,
		write => $COMMAND_WRITE,
	);


	# If there already is an instance running, this will return TRUE; there's no
	# race condition because the check is already performed at construction time.
	if ($app->is_running) {
		my $data = [$text, '/etc/passwd'];
		my $response = $app->send_message($COMMAND_WRITE, data => '/etc/passwd');
		return 0;
	}


	# Create the single application instance and wait for other requests
	my $window = create_application($app, $text);
	Gtk2->main();
	
	return 0;
}


#
# Called when the application needs to be created. This happens when there's no
# other instance running.
#
sub create_application {
	my ($app, $text) = @_;

	# Standard window and windgets
	my $window = Gtk2::Window->new();
	$window->set_title("Unique - Example");
	$window->set_size_request(480, 240);
	my $textview = Gtk2::TextView->new();
	my $scroll = Gtk2::ScrolledWindow->new();
	my $buffer = $textview->get_buffer;

	$buffer->insert($buffer->get_end_iter, "$text\n");

	# Widget packing
	$scroll->add($textview);
	$window->add($scroll);
	$window->show_all();

	# Widget signals
	$window->signal_connect(delete_event => sub {
		Gtk2->main_quit();
		return TRUE;
	});

	# Listen for new commands
	$app->watch_window($window);
	$app->signal_connect('message-received' => sub {
		my ($app, $command, $message, $time) = @_;
		
		my $text = Dumper($message->get);
		$buffer->insert($buffer->get_end_iter, "$text\n");
		
		# Must return a "Gtk2::UniqueResponse"
		return 'ok';
	});

	return $window;
}


lib/Gtk2/Unique.pm
package Gtk2::Unique;

=head1 NAME

Gtk2::Unique - Use single instance applications

=head1 SYNOPSIS

	use Gtk2 '-init';
	use Gtk2::Unique;
	
	my $COMMAND_FOO = 1;
	my $COMMAND_BAR = 2;
	
	my $app = Gtk2::UniqueApp->new(
		"org.example.UnitTets", undef,
		foo => $COMMAND_FOO,
		bar => $COMMAND_BAR,
	);
	
	
	if ($app->is_running) {
		# The application is already running, send it a message
		my ($text) = @ARGV ? @ARGV : ("Foo text here");
		$app->send_message_by_name('foo', text => $text);
	}
	else {
		# Create the single application instance and wait for other requests
		my $window = create_application_window($app);
		Gtk2->main();
	}
	
	
	sub create_application_window {
		my ($app) = @_;
		
		my $window = Gtk2::Window->new();
		my $label = Gtk2::Label->new("Waiting for a message");
		$window->add($label);
		$window->set_size_request(480, 120);
		$window->show_all();
		
		$window->signal_connect(delete_event => sub {
			Gtk2->main_quit();
			return TRUE;
		});
		
		# Watch the main window and register a handler that will be called each time
		# that there's a new message.
		$app->watch_window($window);
		$app->signal_connect('message-received' => sub {
			my ($app, $command, $message, $time) = @_;
			$label->set_text($message->get_text);
			return 'ok';
		});
	}

=head1 DESCRIPTION

Gtk2::Unique is a Perl binding for the C library libunique which provides a
way for writing single instance application. If you launch a single instance
application twice, the second instance will either just quit or will send a
message to the running instance.

For more information about libunique see:
L<http://live.gnome.org/LibUnique>.

=head1 BUGS & API

This is the first release of the module, some bugs can be expected to be found.
Furthermore, the Perl API is not yet frozen, if you would like to suggest some
changes please do so as fast as possible.

=head1 AUTHORS

Emmanuel Rodriguez E<lt>potyl@cpan.orgE<gt>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Emmanuel Rodriguez.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut

use warnings;
use strict;
use base 'DynaLoader';

use Gtk2;

our $VERSION = '0.04';

sub dl_load_flags { $^O eq 'darwin' ? 0x00 : 0x01 }

__PACKAGE__->bootstrap($VERSION);

1;

